proposals/trusted_registry_crd.md

```diff
@@ -0,0 +1,128 @@
+  images from private Harbor repository accessible at https://core.harbor.domain/library/nginx:1.16, but deny pull
+  requests to the core.harbor.domain/library/redis:5 or https://docker.io/library/nginx:1.16 repositories.
+- Statically scan container images for critical, high or medium vulnerabilities. In general, these rules are configurable.
+- Use a Kubernetes admission controller to deny workloads with non-compliant container images. For example, one may
+  allow `kubectl run nginx-pod --image=core.harbor.domain/library/nginx:1.16`, but deny
+  `kubectl run redis-pod --image core.harbor.domain/library/redis:5`.
+
+## Problem statement
+
+- Starboard CLI and Starboard Security Operator are able to scan workloads already deployed in a cluster. Hence, an 
+  admission controller cannot use vulnerability reports for compliance checks.
+- The `vulnerabilities.aquasecurity.github.io` resources are scoped to the same namespace as the underlying workload.
+  However, for admission controllers we might want to introduce cluster-scoped resource to represent vulnerability
+  report associated with a given container image digest, for example `registryvulnerabilities.aquasecurity.github.io`.
+- Container images are not first-class citizens in Kubernetes-native world. There’s no built-in resource that represents
+  a container image. Also, there’s no built-in representation of a (trusted) registry.
+- An admission controller has certain performance requirements. Usually we cannot trigger a scanner when the admission
+  controller is invoked for a given Kubernetes API request. Most likely it will take way too long to perform such
+  ad hoc scans.
+
+## Solutions
+
+### Define registries.aquasecurity.github.io resource
+
+First off, introduce a custom resource to represent a (trusted) container registry. For example, we can represent
+a private Harbor registry with the following Kubernetes object:
+
+```yaml
+apiVersion: aquasecurity.github.io/v1alpha1
+kind: Registry
+metadata:
+  name: my-harbor-registry
+spec:
+  name: My Harbor Registry
+  description: |
+    That's the only registry authorized to run production workloads.
+  server: core.harbor.domain
+  imagePullSecrets:
+  - name: my-harbor-registry-credentials
+  insecure: false
+  trustedRepositories:
+  - name: library/nginx:1.16
+  - name: library/alpine
+  - namePattern: trusted/**
+status:
+  trustedRepositoryStatuses:
+  - name: library/nginx
+    scanned: true
+    lastScanTime: 2020-01-01T00:00:00Z
+  - name: library/alpine
+    scanned: false
+    lastScanTime: null
+```
+
+The `Registry` entity has the `spec` property, which allows specifying trusted repositories. The `status` filed might
+contain scan status for individual repositories. In the example above, you can see that the `library/nginx` repository
+has already been scanned, whereas the `library/alpine` repository has not.
```

Comment:
What if the repo is part-scanned? Some image versions might be scanned, others might not...

Reply:
I think that eventually an admission controller might be checking existence of the registryvulnerabilities reports for the given deployment request. The status at the registry level would be just a summary / current status of the reconciliation loop.